src/main/kotlin/com/stefan_grafberger/streamdq/checks/row/RowValueIsNonNegative.kt

```diff
@@ -0,0 +1,68 @@
+/**
+ * Licensed to the University of Amsterdam (UvA) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The UvA licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.stefan_grafberger.streamdq.checks.row
+
+import com.stefan_grafberger.streamdq.checks.RowLevelConstraintResult
+import com.stefan_grafberger.streamdq.checks.TypeQueryableRowMapFunction
+import org.apache.flink.api.common.ExecutionConfig
+import org.apache.flink.api.common.typeinfo.TypeInformation
+import org.apache.flink.streaming.util.typeutils.FieldAccessor
+import org.apache.flink.streaming.util.typeutils.NullCheckingFieldAccessorFactory
+import java.math.BigDecimal
+
+data class RowValueIsNonNegative(
+        override val keyExpressionString: String
+) :
+        RowLevelConstraint() {
+    override fun <T> getCheckResultMapper(
+            streamObjectTypeInfo: TypeInformation<T>,
+            config: ExecutionConfig?
+    ): TypeQueryableRowMapFunction<T> {
+        return RowValueIsNonNegativeRowMapFunction(
+                streamObjectTypeInfo,
+                toString(),
+                keyExpressionString, config
+        )
+    }
+}
+
+class RowValueIsNonNegativeRowMapFunction<T>(
+        streamObjectTypeInfo: TypeInformation<T>,
+        val checkName: String,
+        val keyExpressionString: String,
+        val config: ExecutionConfig?
+) : TypeQueryableRowMapFunction<T>() {
+
+    private val fieldAccessor: FieldAccessor<T, Any>
+
+    init {
+        fieldAccessor = NullCheckingFieldAccessorFactory.getAccessor(streamObjectTypeInfo, keyExpressionString, config)
+    }
+
+    override fun map(value: T): RowLevelConstraintResult {
+        val fieldValue = this.fieldAccessor.get(value)
+        val bigDecimalFieldValue = if (fieldValue is Number) {
+            fieldValue.toDouble().toBigDecimal()
+        } else {
+            throw NotImplementedError("TODO: Support more comparable types")
+        }
+        val notLessThan: (BigDecimal, BigDecimal) -> Boolean = { x: BigDecimal, y: BigDecimal -> x >= y }
```

Comment:
Did not really think this through, but this looks like it might be unnecessarily complicated and I am not sure if it is correct. In the new test case you added, it also doesn’t look like it explicitly constructs test data with a BigDecimal field that contains a BigDecimal that is larger than the supported Double range. Instead of doing all this here, I think you can just update your „if“ in L59 with an additional „if fieldValue is BigDecimal“, and only if it isn‘t, you go into the current „if fieldValue is Number“ and the else branch.
Does that make sense?

Reply:
Hi Stefan, this sounds a good approach, but when I try pass in a list like this `val doubleList = listOf(Double.MAX_VALUE, -Double.MAX_VALUE, 0.0)`, I found the result is wrong.. because in the Number of branch, I convert field by `fieldValue.toDouble().toBigDecimal()` but it is already a Double.

To resolve this, I used a when at L59 and added three conditions, bigDecimal, Double, and other Numbers. And for now it works well with Double.MAX_VALUE

Reply:
pushed this commit https://github.com/stefan-grafberger/StreamDQ/pull/5/commits/e58ea0349d6b5318f46320eee96c5682c090adab